table.go

```diff
@@ -793,9 +800,140 @@ func addPartToGranule(granules []*Granule, p *Part) error {
 // abort a compaction transaction.
 func (t *TableBlock) abort(commit func(), granule *Granule) {
 	for {
-		if granule.pruned.CAS(1, 0) { // unmark pruned, so that we can compact it in the future
+		if granule.metadata.pruned.CAS(1, 0) { // unmark pruned, so that we can compact it in the future
 			commit()
 			return
 		}
 	}
 }
+
+// filterGranule returns false if this granule does not contain useful data.
+func filterGranule(logger log.Logger, filterExpr logicalplan.Expr, g *Granule) bool {
+	if filterExpr == nil {
+		return true
+	}
+
+	switch expr := filterExpr.(type) {
+	default: // unsupported filter
+		level.Info(logger).Log("msg", "unsupported filter")
+		return true
+	case logicalplan.BinaryExpr:
+
+		var min, max *parquet.Value
+		var v scalar.Scalar
+		var leftresult bool
+		switch left := expr.Left.(type) {
+		case logicalplan.BinaryExpr:
+			leftresult = filterGranule(logger, left, g)
+		case logicalplan.Column:
+			var found bool
+			min, max, found = findColumnValues(left.ColumnsUsed(), g)
+			if !found {
+				// If we fallthrough to here, than we didn't find any columns that match so we can skip this granule
+				return false
+			}
+		case logicalplan.LiteralExpr:
+			switch left.Value.(type) {
+			case *scalar.Int64:
+				v = left.Value.(*scalar.Int64)
+			case *scalar.String:
+				v = left.Value.(*scalar.String)
+			}
+		}
+
+		switch right := expr.Right.(type) {
+		case logicalplan.BinaryExpr:
+			switch expr.Op {
+			case logicalplan.AndOp:
+				rightresult := filterGranule(logger, right, g)
```

Comment:
This can be skipped if leftresult == false

Reply:
You want to send a PR? (I know Thor is offline for the next few days)

Reply:
https://github.com/polarsignals/arcticdb/pull/52 Opened.